subprojects/core/src/main/java/org/gradle/internal/classpath/types/ExternalPluginsInstrumentationTypeRegistry.java

```diff
@@ -51,21 +49,26 @@ public Set<String> getSuperTypes(String type) {
         if (type.startsWith(GRADLE_CORE_PACKAGE_PREFIX)) {
             superTypes = gradleCoreInstrumentationTypeRegistry.getSuperTypes(type);
         }
-        return !superTypes.isEmpty() ? superTypes : computeAndCacheSuperTypesForExternalType(type);
+        return !superTypes.isEmpty() ? superTypes : Collections.unmodifiableSet(computeAndCacheSuperTypesForExternalType(type));
     }
 
     private Set<String> computeAndCacheSuperTypesForExternalType(String type) {
-        return Collections.unmodifiableSet(this.instrumentedSuperTypes.computeIfAbsent(type, this::computeInstrumentedSuperTypes));
+        // We intentionally avoid using computeIfAbsent, since ConcurrentHashMap doesn't allow map modifications in the computeIfAbsent map function.
+        // The result is, that multiple threads could recalculate the same key multiple times, but that is not a problem in our case.
+        Set<String> computedInstrumentedSuperTypes = instrumentedSuperTypes.get(type);
+        if (computedInstrumentedSuperTypes == null) {
+            computedInstrumentedSuperTypes = computeInstrumentedSuperTypes(type);
+            instrumentedSuperTypes.put(type, computedInstrumentedSuperTypes);
+        }
+        return computedInstrumentedSuperTypes;
```

Comment:
❓ I don't understand how this is different to using `computeIfAbsent()`. What'd break if we went back to using `computeIfAbsent()`?

Reply:
We recursively update the map. But that is not allowed for `ConcurrentHashMap` and it throws exception.

See: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-

```
...so the computation should be short and simple, and must not attempt to update any other mappings of this map.
```